037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+
+```json
+{
+  "object": {
+    "ref": "e4be0b367d7bd34580f4842dd09e7b59b6097b25"
+  }
+}
+{
+  "object": {
+    "ref": "5a052ba6438d754f73252283c6b6429f2a74dbff"
+  }
+}
+```
+
+Per the resource interface, these object properties would be saved as versions
+of the resource in the order returned by `check`.
+
+When it comes time to fetch a version in a build, the `git` prototype would
+receive a `get` message with the following object:
+
+```json
+{
+  "object": {
+    "uri": "https://github.com/concourse/rfcs",
+    "branch": "master",
+    "ref": "e4be0b367d7bd34580f4842dd09e7b59b6097b25"
+  }
+}
+```
+
+Note that the object properties have been merged into the original set of
+properties.
+
+## Pipeline Usage
+
+A new `run` step type will be introduced. It takes the name of a message to
+send, and a `type` denoting the prototype to use:
+
+```yaml
+prototypes:
+- name: some-prototype
+  type: registry-image
+  source:
+    repository: vito/some-prototype
+
+jobs:
+- name: run-prototype
+  plan:
+  - run: some-message
+    type: some-prototype
+```
+
+So far this example is not very useful. There aren't many things I can think to
+run that don't take any inputs or configuration or produce any outputs. Let's
+try giving it an input and collecting its resulting output. This can be done
+with `inputs`/`input_mapping` and `outputs`/`output_mapping`:
+
+```yaml
+prototypes:
+- name: oci-image
+  type: registry-image
+  source:
+    repository: vito/oci-image-prototype
+
+resources:
+- name: my-image-src
+  type: git
+  source:
+    uri: https://example.com/my-image-src
+
+- name: my-image
+  type: registry-image
+  source:
+    repository: example/my-image
+    username: some-username
+    password: some-password
+
+jobs:
+- name: build-and-push
+  plan:
+  - get: my-image-src
+  - run: build
+    type: oci-image
+    input_mapping: {context: my-image-src}
+    outputs: [image]
+  - put: my-image
+    params: {image: image/image.tar}
+```
+
+This example also makes use of resources, which are also implemented as
+prototypes with special pipeline pipeline semantics and step syntax. These
+"resource prototypes" are described in [RFC #38][rfc-38].
+
+
+## Out of scope
+
+* [Richer metadata](https://github.com/concourse/concourse/issues/310) - this
+  hasn't gained much traction and probably needs more investigation before it
+  can be incorporated. This should be easy enough to add as a later RFC.
+
+
+## Etymology
+
+* The root of the name comes from [Prototype-based
+  programming](https://en.wikipedia.org/wiki/Prototype-based_programming), a
+  flavor of object-oriented programming, which the concepts introduced in this
+  proposal mirror somewhat closely. There are also some hints of actor-based
+  programming - actors handling messages and emitting more actors - but that
+  felt like more of a stretch.
+
+  This is also an opportunity to frame Concourse as "object-oriented CI/CD," if
+  we want to.
```

Comment:
It's true that Resources have helped reason about Concourse all as one system, but just as there is more to Haskell than just the IO monad (or even Monads in general), there is more to what people are doing with Concourse than what Resources are meant to be used for (see [What makes a resource?](https://blog.concourse-ci.org/reinventing-resource-types/)).

With the advent of this proposal I expect dashboards and pipeline UIs to become more precise. Today we overload Resources beyond essential pipeline metaphors like artifacts and dependencies with other meanings like notifications, deployments, sets of values (e.g. the set of all branches/pull requests). If you cross your eyes and hold your breath you can get these concepts to "almost" be a resource - until they aren't, which is where we've been stuck with a local maxima with user pain ranging from mild (UI cluttered with notifications) to acute (PRs are either hard, through pipeline automation, or easy but with pitfalls).

Fitting everything into one shape worsens the signal-to-noise ratio and means I have to *think* about each object on my screen to understand what it actually is, as they otherwise are shown the same - is this output actually an artifact being delivered, or is it just another resource being used to send a Slack message? Do Slack messages really justify taking up a good chunk of my pipeline UI?

What we're discussing here is a protocol that can support more than just the 'resources' idea - it should not diminish it, it should strengthen what remains, and allow these other concepts be framed properly.

Does that help? 🤔